csharp/src/Ice/IncomingFrame.cs

```diff
@@ -61,14 +96,15 @@ public void DecompressPayload()
 
                 Debug.Assert(Payload.Array == Encapsulation.Array);
 
-                byte[] decompressedData = new byte[Encapsulation.Offset + sizeLength + decompressedSize];
-                Encapsulation.Array.AsSpan(0, Encapsulation.Offset + offset).CopyTo(decompressedData);
+                byte[] decompressedData = new byte[Payload.Count - size + decompressedSize];
+                int encapsulationOffset = Encapsulation.Offset - Payload.Offset;
```

Comment:
`encapsulationOffset = Encapsulation.Offset - Payload.Offset;` ... pretty confusing.

Reply:
Encapsulation is a subset of Payload, and this computes the offset of the encapsulation in the payload segment, so we can copy all the bytes from payload up to the start of the encapsulation to the decompressed data.

Encapsulation.Offset and  Payload.Offset are absolute offsets on the underlying array and it can be displaced, we create the payload by slicing the incoming buffer.

This is a bit more complex now because the binary context is write in Payload after the  encaps and we have to copy it around when decompress the encaps

Reply:
Reworked this a bit and replace it with `gzipOffset` the offset to the start of the gzip data.